miner/payload_building.go

```diff
@@ -0,0 +1,160 @@
+// which has an empty transaction set and then keep update it in order to maximize
+// the revenue. Therefore, the empty-block here is always available and full-block
+// will be set/updated afterwards.
+type Payload struct {
+	empty *types.Block
+	full  *types.Block
+	stop  chan struct{}
+	lock  *sync.RWMutex
+	cond  *sync.Cond
+}
+
+// newPayload initializes the payload object.
+func newPayload() *Payload {
+	lock := new(sync.RWMutex)
+	return &Payload{
+		stop: make(chan struct{}),
+		lock: lock,
+		cond: sync.NewCond(lock),
+	}
+}
+
+// update updates the full-block with latest built version.
+func (payload *Payload) update(block *types.Block) {
+	payload.lock.Lock()
+	defer payload.lock.Unlock()
+
+	select {
+	case <-payload.stop:
+		return // reject stale update
+	default:
+	}
+	payload.full = block
+	payload.cond.Broadcast() // fire signal for notifying full block
+}
+
+// Resolve returns the latest built payload and also terminates the background
+// thread for updating payload. It's safe to be called multiple times.
+func (payload *Payload) Resolve() *beacon.ExecutableDataV1 {
+	payload.lock.Lock()
+	defer payload.lock.Unlock()
+
+	select {
+	case <-payload.stop:
+	default:
+		close(payload.stop)
+	}
+	if payload.full != nil {
+		return beacon.BlockToExecutableData(payload.full)
+	}
+	return beacon.BlockToExecutableData(payload.empty)
+}
+
+// ResolveEmpty is basically identical to Resolve, but it expects empty block only.
+// It's only used in tests.
+func (payload *Payload) ResolveEmpty() *beacon.ExecutableDataV1 {
+	payload.lock.Lock()
+	defer payload.lock.Unlock()
+
+	return beacon.BlockToExecutableData(payload.empty)
+}
+
+// ResolveFull is basically identical to Resolve, but it expects full block only.
+// It's only used in tests.
+func (payload *Payload) ResolveFull() *beacon.ExecutableDataV1 {
+	payload.lock.Lock()
+	defer payload.lock.Unlock()
+
+	if payload.full == nil {
+		select {
+		case <-payload.stop:
+			return nil
+		default:
+		}
+		payload.cond.Wait()
+	}
+	return beacon.BlockToExecutableData(payload.full)
+}
+
+// buildPayload builds the payload according to the provided parameters.
+func (w *worker) buildPayload(args *BuildPayloadArgs) (*Payload, error) {
+	// Build the initial version with no transaction included. It should be fast
+	// enough to run. The empty payload can at least make sure there is something
+	// to deliver for not missing slot.
+	empty, err := w.getSealingBlock(args.Parent, args.Timestamp, args.FeeRecipient, args.Random, true)
+	if err != nil {
+		return nil, err
+	}
+	// Construct a payload object for return.
+	payload := newPayload()
+	payload.empty = empty
+
+	// Spin up a routine for updating the payload in background. This strategy
+	// can maximum the revenue for including transactions with highest fee.
+	go func() {
+		// Setup the timer for re-building the payload. The initial clock is kept
+		// for triggering process immediately.
+		timer := time.NewTimer(0)
```

Comment:
Not sure I follow. Why do you want to trigger immediately?

Reply:
Is there any reason to believe that the block we get when we interrupt immediately is any different from the `empty` block?

Reply:
Ah, right. So the answer is yes, even if we interrupt immediately, the miner internally was given `newpayloadTimeout` (2s) allowance in which to add transactions.

Reply:
Why the block will be interrupted immediately?

Here we capture the first signal from timer to build the *first full block*. Internally miner will give it 2 seconds to fulfill the block. Once the full block is built, then we reset the timer for another round(with the interval specified by `w.recommit`).